### PR TITLE
Re-added the Turbine

### PR DESCRIPTION
**Pull request by Generalcamo**
image

This is intended for future use.

This is a direct port from Paradise, modified to work with our code.

### Asset Licenses
The following assets that **have not** been created by myself are included in this PR:

| Path | Original Author | License |
| --- | --- | --- |
| sound/effects/electheart.ogg | tigercat2000 (Paradise) | CC-BY-SA |
| icons/obj/pipeturbine.dmi | [PestoVerde322](https://github.com/PestoVerde322) (Beestation) | CC-BY-SA |

### CODE REVIEW

**Comment by Generalcamo** [comment]
!review

File: code/modules/power/turbine.dm [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+	starter = FALSE
+	last_overheat = world.time
+	overheat -= 50
+	radio.autosay(OVERHEAT_MESSAGE, name, "Engineering")
+	playsound(src, 'sound/machines/buzz-two.ogg', 100, FALSE, 40, 30, falloff_distance = 10)
+
+/obj/machinery/power/compressor/proc/time_until_overheat_done()
+	return max(last_overheat + OVERHEAT_TIME - world.time, 0)
+
+/obj/machinery/power/compressor/process()
```
Comment (FluffyGhoster) [inline code/modules/power/turbine.dm:163]: process needs the parameter `ticks_per_seconds` as parent and to multiply everything related to time with it

File: code/modules/power/turbine.dm [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+	productivity = P / 6
+
+/obj/machinery/power/turbine/proc/locate_machinery()
+	if(compressor)
+		return
+	compressor = locate() in get_step(src, get_dir(outturf, src))
+	if(compressor)
+		compressor.locate_machinery()
+
+/obj/machinery/power/turbine/process()
```
Comment (FluffyGhoster) [inline code/modules/power/turbine.dm:243]: process needs the parameter `ticks_per_seconds` as parent and to multiply everything related to time with it

File: code/modules/power/turbine.dm [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+				to_chat(usr, SPAN_ALERT("The turbine is overheating, please wait [time_until_done / 10] seconds for cooldown procedures to complete."))
+				playsound(src, 'sound/effects/electheart.ogg', 100, FALSE, 40, 30, falloff_distance = 10)
+				. = TRUE
+			else if(compressor?.turbine)
+				if(!compressor.starter)
+					playsound(compressor, 'sound/mecha/powerup.ogg', 100, FALSE, 40, 30, falloff_distance = 10)
+				compressor.starter = !compressor.starter
+				. = TRUE
+
+/obj/machinery/computer/terminal/turbine_computer/process()
```
Comment (FluffyGhoster) [inline code/modules/power/turbine.dm:415]: process needs the parameter `ticks_per_seconds`

File: code/modules/power/turbine.dm [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+	var/obj/machinery/power/turbine/turbine
+	var/datum/gas_mixture/gas_contained
+	var/turf/simulated/inturf
+	var/starter = FALSE
+	var/rpm = 0
+	var/rpm_threshold = NONE
+	var/rpmtarget = 0
+	var/capacity = 1e6
+	var/comp_id = 0
+	var/efficiency
```
Comment (FluffyGhoster) [inline code/modules/power/turbine.dm:50]: DMDocs

File: code/modules/power/turbine.dm [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+	name = "gas turbine generator"
+	desc = "A gas turbine used for backup power generation."
+	icon = 'icons/obj/pipeturbine.dmi'
+	icon_state = "turbine"
+	anchored = TRUE
+	density = TRUE
+	var/opened = FALSE
+	var/obj/machinery/power/compressor/compressor
+	var/turf/simulated/outturf
+	var/lastgen
```
Comment (FluffyGhoster) [inline code/modules/power/turbine.dm:73]: DMDocs

File: code/modules/power/turbine.dm [hunk 6]
```diff
@@ -0,0 +1,431 @@
...
+	/// If the turbine is outputing enough to visibly affect its sprite
+	var/generator_threshold = FALSE
+	var/productivity = 1
```
Comment (FluffyGhoster) [inline code/modules/power/turbine.dm:76]: DMDoc

File: code/modules/power/turbine.dm [hunk 7]
```diff
@@ -0,0 +1,431 @@
...
+
+/obj/machinery/computer/terminal/turbine_computer
+	name = "gas turbine control computer"
+	desc = "A computer to remotely control a gas turbine. Link it to a turbine via use of a multitool."
+	icon_screen = "turbinecomp"
+	icon_keyboard = "tech_key"
+	var/obj/machinery/power/compressor/compressor
+	var/list/obj/machinery/door/blast/doors
+	var/id = 0
+	var/door_status = 0
```
Comment (FluffyGhoster) [inline code/modules/power/turbine.dm:90]: DMDocs

**Comment by furrycactus** [comment]
While I'm still not sold on this replacing #19767 due to the power production vs. demand on the Horizon, after some thought I just realised this could be absolutely fantastic for several ghost role off-ships who tend to have a much lower power draw. It's a tiny bit more involved than a port gen or miniature reactor while not being overly complicated, but not so large and obtrusive that it takes up as much space as an SM, and it doesn't drastically generate far more power than is needed.

This could be really great as it can give off-ship engineer roles something more to actually do instead of being a glorified graphite/tritium sheet loader, while not alienating people who aren't super skilled with Engineering/Atmospherics tasks who do actually prefer just using the port gens.

**Comment by FluffyGhoster** [comment]
Also, this shouldn't be mapped into runtime; runtime is made to load as fast as possible, and should only contains the minimum to test the essentials at that scope, the turbine can be spawned if it needs to be tested, or placed for some tests

**Review by Arrow768 — APPROVED** [review]

**Review by NonQueueingMatt — APPROVED** [review]

**Pull request merged**